### DISCUSSION TITLE
How to create multiple idependent route

### DISCUSSION THREAD

**Discussion by azzamsa** [poem-web/poem#192] [Q&A] [answered]
I have a web app act as both auth server and middleware to route to other service.

1. User can register themselves using `/graphql` endpoint (graphql output)
2. After registration, user are able to access `/other-service-endpoint/feature` (json output)

Without the first step, my app is simply:

```rust
 let app = Route::new()
        .at("/", get(routes::graphql_playground))
        .at("/graphql", post(routes::graphql_handler))
        .data(schema);
```

Now, I want to be able for every user accessing `/other-service-endpoint/foo` or `/other-service-endpoint/bar` to route them into `other-service-ip/foo` and `other-service-ip/bar` respectively`

I was thinking this is suitable for middleware, but I go with most easiest path. Instead of using `/other-service-endpoint/bar` as url. I ask client to acces `/other-service-endpoint/` and set `bar` in header request.

---

How to create mutiple route for this?

Something like?

```rust
    let app = Route::multiple(
        Route::new()
            .at("/graphql", post(routes::graphql_handler))
            .data(schema),
        Route::new().at("/other-service", post(routes::route_to_other_service)),
    );
```

Or, do you have any better approach?

Thanks for Poem.

**Comment by sunli829** [comment]
It looks like what you need is `nest`?

```rust
let app = Route::new()
    .at("/graphql", post(routes::graphql_handler))
    .nest("/other-service", post(routes::route_to_other_service))
    .data(schema);
```

**Comment by azzamsa** [comment]
I choose to go with `multiple data` instead of separating them. And accessing specific data using type matching in the handler.

```rust
// main.rs

let app = Route::new()
        .at("/", get(routes::graphql_playground))
        .at("/graphql", post(routes::graphql_handler))
        .at("/other/*path", get(routes::other).post(routes::other))
        .data(schema)
        .data(Arc::clone(&server_context));
 ```
 
 ```rust
 // routes.rs

#[handler]
pub async fn graphql_handler(schema: Data<&AppSchema>, req: Json<Request>) -> Json<Response> {
    Json(schema.execute(req.0).await)
}

#[handler]
pub async fn risetai(
    ctx: Data<Arc<ServerContext>>,
    Path(path): Path<String>,
    body: Json<Value>,
    headers: &HeaderMap,
) -> String {
    ctx.other_service.route(path, body, headers).await
}
 ```
 
 However, I got an error:

```rust
error[E0277]: the trait bound `poem::web::Data<Arc<ServerContext>>: FromRequest<'_>` is not satisfied
  --> src/routes.rs:26:1
   |
26 | #[handler]
   | ^^^^^^^^^^ the trait `FromRequest<'_>` is not implemented for `poem::web::Data<Arc<ServerContext>>`
   |
   = help: the following implementations were found:
             <poem::web::Data<&'a T> as FromRequest<'a>>
   = note: this error originates in the attribute macro `handler` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `app` due to previous error
```

I think it is doable in [actix-web](https://github.com/skerkour/bloom-legacy/blob/5c94122222b4b99e79f4abd7bd7946d42e2b9547/bloom/http_server/src/api/files/commands/complete_file_upload.rs#L12)

Is using `Arc<T>` not supported in Poem handler?

**Comment by sunli829** [comment, marked as answer]
```
#[handler]
pub async fn risetai(
   ctx: Data<&Arc<ServerContext>>,  //   <<<<<  & reference is required
   Path(path): Path<String>,
   body: Json<Value>,
   headers: &HeaderMap,
) -> String {
   ctx.other_service.route(path, body, headers).await
}
```

**Reply by azzamsa** [reply]
Worked like a charm!

Thanks a lot for taking a time to reply fast.
